CDT/include/Triangulation.hpp

```diff
@@ -1262,6 +1266,282 @@ bool Triangulation<T, TNearPointLocator>::isFlipNeeded(
+{
+    EdgeQue encroachedEdges;
+    for(EdgeUSet::const_iterator cit = fixedEdges.begin();
+        cit != fixedEdges.end();
+        ++cit)
+    {
+        const Edge edge = *cit;
+        if(isEncroached(v, edge))
+        {
+            encroachedEdges.push(edge);
+        }
+    }
+    return encroachedEdges;
+}
+
+template <typename T, typename TNearPointLocator>
+std::pair<TriIndVec, IndexSizeType>
+Triangulation<T, TNearPointLocator>::resolveEncroachedEdges(
+    EdgeQue encroachedEdges,
+    const V2d<T>& v,
+    const bool validV,
+    const bool fillBadTriangles,
+    const RefineTriangles::Enum refinementConstrain,
+    const T badTriangleThreshold)
+{
+    IndexSizeType numOfSplits = 0;
+    std::vector<TriInd> badTriangles;
+
+    while(!encroachedEdges.empty() && numOfSteinerPoints < maxSteinerPoints)
+    {
+        Edge edge = encroachedEdges.front();
+        encroachedEdges.pop();
+        if(!hasEdge(edge.v1(), edge.v2()))
+        {
+            continue;
+        }
+        TriInd iT = edgeTriangle(edge);
+        const Triangle& t = triangles[iT];
+        VertInd i = splitEncroachedEdge(
+            edge, iT, edgeNeighbor(triangles[iT], edge.v1(), edge.v2()));
+        ++numOfSplits;
+
+        TriInd start = m_vertTris[i];
+        TriInd currTri = start;
+        do
+        {
+            const Triangle& t = triangles[currTri];
+            if(fillBadTriangles &&
+               isBadTriangle(t, refinementConstrain, badTriangleThreshold))
+            {
+                badTriangles.push_back(currTri);
+            }
+            for(int i = 0; i < 3; ++i)
+            {
+                const Edge edge(t.vertices[i], t.vertices[cw(i)]);
+                if(fixedEdges.find(edge) == fixedEdges.end())
+                    continue;
+                const TriInd iT = currTri;
+                const TriInd iTopo =
+                    edgeNeighbor(triangles[iT], edge.v1(), edge.v2());
+                const Triangle& tOpo = triangles[iTopo];
+                VertInd v1 = opposedVertex(t, iTopo);
+                VertInd v2 = opposedVertex(tOpo, iT);
+                if(isEncroached(vertices[v1], edge) ||
+                   isEncroached(vertices[v2], edge) ||
+                   (validV && isEncroached(v, edge)))
+                {
+                    encroachedEdges.push(edge);
+                }
+            }
+            currTri = t.next(i).first;
+        } while(currTri != start);
+    }
+    return std::make_pair(badTriangles, numOfSplits);
+}
+
+template <typename T, typename TNearPointLocator>
+VertInd Triangulation<T, TNearPointLocator>::splitEncroachedEdge(
+    const Edge splitEdge,
+    const TriInd iT,
+    const TriInd iTopo)
+{
+    const VertInd iMid = static_cast<VertInd>(vertices.size());
+    const V2d<T>& start = vertices[splitEdge.v1()];
+    const V2d<T>& end = vertices[splitEdge.v2()];
+    T split = T(0.5);
+    if(isSteinerVertex[splitEdge.v1()] || isSteinerVertex[splitEdge.v1()])
```

Comment:
Should this be instead?
```suggestion
    if(isSteinerVertex[splitEdge.v1()] || isSteinerVertex[splitEdge.v2()])
```